Add Env0 join method

timothyb89 opened the pull request:
This adds a new `env0` join method to support joining from env0 workflows, and especially the embedded tbot in the Terraform provider when running on env0.

This is the first OIDC join method on the new join service, and includes new messages that will be shared among all OIDC join methods. Unlike legacy OIDC join methods which will be ported to the new join service, this new join method will not support legacy joining and won't be backported beyond v18.

Closes https://github.com/gravitational/teleport/issues/53798

changelog: Add new `env0` join method to support joining within Env0 workflows

timothyb89 committed:
Add Env0 method

This adds a new `env0` join method to support joining from env0
workflows, and especially the embedded tbot in the Terraform provider
when running on env0.

This is the first OIDC join method on the new join service, and will
not support legacy joining. As such, it won't be backported beyond
v18.

Closes #53798

changelog: Add new `env0` join method to support joining within Env0 workflows

timothyb89 committed:
Fix imports

timothyb89 committed:
Fix failing test

timothyb89 committed:
Add env0 token source

timothyb89 committed:
Fix missing env0 join method in tbot

timothyb89 committed:
Fix env0 validation

This fixes some issues in env0 token validation:
- azp check is disabled since it's set to a random (ish) value
- Use correct audience URL (their docs specified the wrong value)

timothyb89 committed:
Fix incorrect date in new file copyright header

timothyb89 commented on lib/join/env0/validator.go:
```diff
@@ -0,0 +1,80 @@
+
+	// env0Audience is the audience for the token. This is unfortunately hard
+	// coded.
+	env0Audience = "https://prod.env0.com"
```
The example token in their public docs unfortunately specify the wrong value here 🙁  https://docs.envzero.com/guides/integrations/oidc-integrations#format-of-the-openid-connect-id-token

(At least the value in the table is correct, but I'm glad I was able to test it explicitly...)

timothyb89 committed:
Fix generated tfschema

timothyb89 committed:
Fix failing test after adding the token source

timothyb89 commented:
I've successfully tested this on a real Env0 tenant by roughly following my old Terraform Cloud steps here: https://gist.github.com/timothyb89/5bb905e8c7065ae0601e83f816af8056

It's unfortunately not super easy to replicate since trial tenants aren't readily available to the public, but I can probably invite anyone to my trial account who really wants to validate themselves.

Actual lines changed is more like 1376, but the bot doesn't exclude a lot of the generated code, ah well.

boxofrad reviewed (approved):
This is great! 👏🏻

ptgott reviewed (commented):
Since this change includes both docs and API changes, we'll want to make sure the docs include the minimum Teleport version that supports the changes (assuming this gets released in a minor version).

ptgott commented:
> Are you asking for the automatically generated docs to include a version of Teleport that said feature will land in - even though that cannot be known ahead of time?

We've had it happen sometimes that API changes land in minor Teleport versions, but since we build the docs based on release branches, the docs appear before we release the change. This has caused customer confusion. There doesn't seem to be a good systematic solution to this, so I'm just spreading awareness.

nklaassen commented on lib/join/server_env0.go:
```diff
@@ -0,0 +1,129 @@
+	}
+
+	// There are no additional checks for the token join method, just make the
+	// result message and return it.
```
```suggestion
	// Make and return the final result message it.
```

timothyb89 replied:
Whoops! I'll just remove the comment outright, it doesn't serve much purpose when it's just doing expected stuff.

nklaassen commented on lib/join/server_env0.go:
```diff
@@ -0,0 +1,129 @@
+}
+
+// handleTokenJoin handles join attempts for the token join method.
+func (s *Server) handleEnv0Join(
```
nit: the comment has copy/paste error

but, I think everything in this method except the token validator and the allow rule checker will be common for all OIDC join methods

what would you think of having a single handleOIDCJoin method that accepts the token validator and allow rule checker as an argument, maybe this signature

```go
// oidcTokenValidator is a function type that validates an OIDC token and checks that
// it matches an allow rule configured in the provision token.
type oidcTokenValidator func(
	ctx context.Context,
	provisionToken types.ProvisionToken,
	idToken []byte,
) (rawClaims any, joinAttrs *workloadidentityv1.JoinAttrs, err error)
```

to reduce duplication and separate the concerns of handling the message stream from actually checking the token

nklaassen commented on lib/join/server_env0_test.go:
```diff
@@ -0,0 +1,535 @@
+	}
+	// isBadParameter := func(t require.TestingT, err error, _ ...any) {
+	// 	require.True(t, trace.IsBadParameter(err), "expected Bad Parameter error, actual error: %v", err)
+	// }
```
```suggestion
```

timothyb89 committed:
First round of code review

timothyb89 committed:
Add example token claims in comment

timothyb89 committed:
Move common OIDC handling logic into a separate handler

Specific validation logic is now separate and the validator is passed
as an argument to a generic `handleOIDCJoin()` handler.

timothyb89 committed:
Merge remote-tracking branch 'origin/master' into timothyb89/env0-joining

timothyb89 committed:
Fix imports

timothyb89 committed:
Rename joinclient/join_env0.go to join_oidc.go

timothyb89 committed:
Merge branch 'master' into timothyb89/env0-joining
